src/FastExpressionCompiler/FastExpressionCompiler.cs

```diff
@@ -2095,14 +2095,16 @@ private static void EmitDefault(Type type, ILGenerator il)
                     --paramIndex;
                 if (paramIndex != -1)
                 {
+                    bool isArgumentByRef = byRefIndex == paramIndex;
+
                     if ((closure.Status & ClosureStatus.ShouldBeStaticMethod) == 0)
                         ++paramIndex; // shift parameter index by one, because the first one will be closure
 
                     closure.LastEmitIsAddress = !paramExpr.IsByRef && paramType.IsValueType() &&
                         ((parent & ParentFlags.InstanceCall) == ParentFlags.InstanceCall || 
                          (parent & ParentFlags.MemberAccess) != 0);
 
-                    if (closure.LastEmitIsAddress)
+                    if (closure.LastEmitIsAddress || (!paramExpr.IsByRef && isArgumentByRef))

```

Comment:
This is slightly convoluted way of checking if we are given non-reference parameter but our argument expects a reference. In this case it should load address and not value. Hoping that this is a right place to do this.